### DISCUSSION TITLE
LinkNeighborLoader temporal sampling based on edge timestamps

### DISCUSSION THREAD

**Discussion by zimengzhou1** [pyg-team/pytorch_geometric#6924] [Q&A] [answered]
From what I understand we can specify `edge_label_time` and `time_attr` so that newly sampled neighbors will have a timestamp less than the edge being sampled on. Is there a way to only sample nodes such that the edges to those nodes have a timestamp less than the timestamp of the edge being sampled on?

**Comment by wsad1** [comment, marked as answer]
Currently we only support a `time_attr` for nodes, edges can't have a time attribute. So currently what you are suggesting isn't possible with Pyg.

**Reply by zimengzhou1** [reply]
Thanks for the answer! I thought `edge_label_time` specifies the timestamp for edge indices. If we wanted to sample from a dynamic graph wouldn't it make sense to only use edges that have a timestamp less than the timestamp of the edge being sampled? Would this be something that could be implemented?

**Reply by rusty1s** [reply]
Yes, we will add support for edge-time dependent sampling soon, but for now, we sadly only support node-time dependent sampling.

**Reply by LMscope** [reply]
Thank you for the question and the answer. 
However, I have a follow up question, if this is the case (node-time dependent sampling) for now, does this mean the **TGN implementation** is considered a `discrete time dynamic` or `continuous time dynamic`?
I appreciate your answer.

**Reply by rusty1s** [reply]
The TGN implementation is considered a continuous time dynamic graph.

**Comment by GianlucaDeStefano** [comment]
Hi @rusty1s, apologies for resurrecting this old post. I am also interested in temporal edge sampling, has there been any update on this front by any chance?

**Reply by rusty1s** [reply]
Should land in about a week or two. Fingers crossed.

**Reply by GianlucaDeStefano** [reply]
That's great! Is the implementation available / being developed on one of the public branches by any chance? 
I would really like to take a look at it. 

In any case, thank you for the amazing work!

**Reply by akihironitta** [reply]
@GianlucaDeStefano It's in progress here: https://github.com/pyg-team/pyg-lib/pull/280, #8372.

**Comment by Jieerbobo** [comment]
Hi @rusty1s, thanks for your excellent work on temporal sampling. Recently I have tried LinkNeighborLoader on a dynamic heterogeneous graph, and I found that the node number in a batch is much larger than the original graph. This only happens when using "time_attr". For static graphs, the node number in a batch is the same with the original graph.

For example, the user number and the item number in the original graph are 16379 and 12577, respectively. But when using LinkNeighborLoader to get batches, both numbers in a batch become larger. I am wondering why this happens? If this means each batch is larger than the original graph? Thanks in advance!

<img width="166" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/95608711/07eb6a04-d534-4688-9e38-4ed718e2d41c">
<img width="220" alt="image" src="https://github.com/pyg-team/pytorch_geometric/assets/95608711/edb3600d-3fb3-4800-b78b-9b5aa76b6323">

train_loader = LinkNeighborLoader(
    data = train_data,
    num_neighbors=[10,5],
    edge_label_index= (("user", "to", "user"), train_data["user", "to", "user"].edge_label_index),
    edge_label=train_data["user", "to", "user"].edge_label,
    edge_label_time=train_data["user", "to", "user"].edge_label_time,
    time_attr="node_time",
    temporal_strategy="last",
    neg_sampling_ratio=0.0,
    batch_size = 256,
    shuffle = True,
)

**Reply by rusty1s** [reply]
If you are using temporal sampling, then each subgraph across seed nodes will be disjoint from any other subgraph (see the `disjoint` option). This is required for temporal sampling in case seed times vary.

**Reply by Linnore** [reply]
Hi, thanks for the PyG's work on supporting temporal sampling! I am confused by how each disjoint subgraph is represented in each batch when using LinkNeighborLoader's temporal sampling.

As I observed, the `disjoint=True` will give `batch.batch` of shape `batch.x.shape[0]`, which indicates the subgraph each node belongs to. But how about `batch.edge_index`? How can we get the edges in each disjoint subgraph?

Is there any documentation / example that can help understanding the workflow of temporal sampling using LinkNeighborLoader?

**Reply by julienberman** [reply]
+1 to this!
